Pkg integration

tegefaulkes opened the pull request:
### Description
This PR is to test using vercel/pkg with native modules.
It imports the following node modules:
- level
- utp-native
- threads.js

It should be able to build via pkg and nix-build method.

### Issues Fixed
* Fixes #27 

### Tasks
- import, build and test level
- import, build and test utp-native
- import, build and test threads

### Final checklist
- import and create hello world example for:
    - [x] level
    - [x] utp-native
    - [x] threads
- build and test via pkg
    - [x] level
    - [x] utp-native
    - [x] threads
- build and test via nix-build
    - [x] level
    - [x] utp-native
    - [x] threads
- [x] modify release.nix to include the required prebuilds in the result.
- [x] final tests of the nix builds. see if they include prebuilds properly.
- [x] test if GH auto release still works.

---

Related issues and PRs:

* https://github.com/vercel/pkg/issues/1264
* https://github.com/vercel/pkg/pull/1273

CMCDragonkai commented on src/bin/worker.ts:
```diff
@@ -0,0 +1,7 @@
+import { expose } from "threads/worker"
```
This should got to `src/worker.ts` or `src/workers/worker.ts` like we do in `js-encryptedfs` or `js-polykey`.

tegefaulkes replied:
moved to `src/lib/workers/worker.ts` and moved the workers test into `src/lib/workers/test-workers.ts`

CMCDragonkai commented on utils.nix:
```diff
@@ -28,10 +28,12 @@ rec {
   ).nodeDependencies;
   node2nixDev = (import (node2nixDrv true) { inherit pkgs nodejs; }).package.overrideAttrs (attrs: {
     src = src;
+    buildInputs = attrs.buildInputs ++ [ nodePackages.node-gyp-build ];
```
Have a check whether both require this override, or just one of `node2nixProd` and `node2nixDev`.

tegefaulkes replied:
Just having it in node2nixProd doesn't work while just having it in node2nixDev does.
Removing it from node2nixProd.

tegefaulkes commented:
Completed most of the checklist, The last thing to do is check of the GH releases works properly with the changes.
Depending on the output we may have to zip the builds now since they are directories instead of just the executable.

CMCDragonkai commented on src/bin/typescript-demo-lib.ts:
```diff
@@ -24,6 +27,50 @@ function main(argv = process.argv): number {
   const sum = nums.num1 + nums.num2;
   process.stdout.write(nums.num1 + ' + ' + nums.num2 + ' = ' + sum + '\n');
 
+  // Level tests.
```
Don't use `console.log` here. It makes the test harder.

Use `process.stdout.write` like above.

Also I'd say put these into `src/lib/...` for each individual functionality. Provider functions like `level.ts` which exports a function that does the below, and same for the other ones.

tegefaulkes replied:
- Replaced each use of `consoel.log()` with `process.stdout.write()`.
- Separated each tests into it's own file inside `src/lib/`

CMCDragonkai commented on release.nix:
```diff
@@ -59,13 +65,16 @@ let
       PKG_IGNORE_TAG = 1;
       buildPhase = ''
         cp ${./package.json} package.json
+        mkdir -p output/prebuilds
```
Does this need `output/prebuilds`? Why not just `./prebuilds`? The `./` is the working directory of the build. As long as this directory doesn't conflict with anything in the `./` which I believe is the produced source files from the upstream derivations.

Remember this is an overridden derivation, the `$src` has already unpacked into the current directory.

tegefaulkes replied:
When I use `./` instead of a temp directory, the result included all of the src files.
I also feel that creating and using a temp directory makes the process more explicit.

CMCDragonkai replied:
Is that because of:

```
--public-packages "*" \
```

Is the `*` picking up everything?

CMCDragonkai replied:
If you must use a a special directory for this. I'd say use `pkg` as the special directory for `pkg` related building.

tegefaulkes replied:
Ok, I can change it to pkg.
as for `--public-packages "*" \` is just flagging everything is a public package since non public packages are included as bytecode by default.. This breaks with the `--no-bytecode` flag.

tegefaulkes replied:
Using pkg 5.3.1 doesn't fix the issue for leveldown prebuilds, neither does adding them as assets.
this seems to be an issue with pkg itself, I've posted an issue at https://github.com/vercel/pkg/issues/1264.

CMCDragonkai commented on release.nix:
```diff
@@ -59,13 +65,16 @@ let
           --no-bytecode \
-          --output out
+          --public-packages "*" \
+          --output output/TypeScript-Demo-Lib
```
Why are we creating a `output` directory?

The `$out` is obviously for the final executable. But I see you're putting `prebuilds` into the `output`. Is that not possible with just the `./` directory?

tegefaulkes replied:
- changing `output` to `pkg`. relevant to above comment.

CMCDragonkai commented on release.nix:
```diff
@@ -15,19 +15,22 @@ let
   buildExe = arch:
     stdenv.mkDerivation rec {
-      name = "${utils.basename}-${version}-win32-${arch}.exe";
+      name = "${utils.basename}-${version}-win32-${arch}";
```
Why change this? The name of the windows executable has an extension of `.exe`.

tegefaulkes replied:
The output is a directory now that includes the .exe

CMCDragonkai replied:
You should leave it as the actual executable. Is there other files necessary for the actual `$out`?

This is not a nix package, so the executable as `$out` is totally fine.

Plus this actually is integrated into the Gitlab CI/CD which expects a single file as `$out`. Changing it to a directory adds extra work to do there.

tegefaulkes replied:
We need to have the output as a directory since we need to include the prebuilds for leveldown.
we can have it as a zip file if needed.

CMCDragonkai replied:
Can this be resolved now back to the way it was?

CMCDragonkai commented:
Is there a way for you test the CI/CD execution of your changes.

You just need to go into the gitlab yaml file and enable this for your branch as well.

Then go to https://gitlab.com/MatrixAI/open-source/typescript-demo-lib/-/pipelines to manually trigger it. Or https://gitlab.com/MatrixAI/open-source/typescript-demo-lib/-/jobs.

CMCDragonkai commented on README.md:
````diff
@@ -117,3 +117,75 @@ npm publish --access public
 git push --tags
 ```
+
+
````
Remove empty line here.

CMCDragonkai commented on README.md:
````diff
@@ -117,3 +117,75 @@ npm publish --access public
 ```
+
+
+### Packaging with vercel.pkg
````
Change this to `### Releasing`.

And fix up the typos in the text. Put it into word/writer if necessary.

CMCDragonkai replied:
Spacing of headers should be:

```
text text test

### Header

text text text
```

CMCDragonkai commented on README.md:
````diff
@@ -117,3 +117,75 @@ npm publish --access public
+  }
+```
+
+#### working with nix-build
````
This should be: `### Native Module Toolchain`.

This section actually not part of `Releasing`.

It should be it's own larger section that comes before `Docs Generation`.

CMCDragonkai commented on README.md:
```diff
@@ -117,3 +117,75 @@ npm publish --access public
+    - (node files)
+
+#### utp-native
+Including utp-native is simpler, you just need to add it as an asset for pkg.
```
Can you explain why `utp-native` is different from `leveldown`?

tegefaulkes replied:
I'm not sure, they include the native bindings almost identically.
leveldown `binding.js`
```
module.exports = require('node-gyp-build')(__dirname)
```
where as utp-native does.
`lib/binding.js`
```
const path = require('path')

module.exports = require('node-gyp-build')(path.join(__dirname, '..'))

```

CMCDragonkai commented on pkgs.nix:
```diff
@@ -1,5 +1,6 @@
   # let rev = "60cce7e5e1fdf62421ef6d4184ee399b46209366"; in
-  let rev = "04a48a2fcd2370f4e8a8c9fd3394531ccdb30f1e"; in
+  # let rev = "04a48a2fcd2370f4e8a8c9fd3394531ccdb30f1e"; in
+  let rev = "53caacaf56640d04180775aee016d2f16d6f083c"; in
```
Which revision is this again? Are we updating this to the latest?

tegefaulkes replied:
I picked the latest pkgs commit when I started on this issue.

CMCDragonkai replied:
You should usually get the commit hash matching the exact merge commit. Which would be this one: https://github.com/NixOS/nixpkgs/commit/3d64b6067f1c457f889e91c59ced273be1d1e4f1

Notice you can see it here:

![image](https://user-images.githubusercontent.com/640797/128017469-1c8828eb-788b-473e-8d3e-d5079707f2af.png)

CMCDragonkai replied:
That's for in the future, which avoids bringing in other confounding variables/changes.

tegefaulkes replied:
Ah, I'll keep that in mind.

CMCDragonkai commented:
> Is there a way for you test the CI/CD execution of your changes.
> 
> You just need to go into the gitlab yaml file and enable this for your branch as well.
> 
> Then go to https://gitlab.com/MatrixAI/open-source/typescript-demo-lib/-/pipelines to manually trigger it. Or https://gitlab.com/MatrixAI/open-source/typescript-demo-lib/-/jobs.

Oh if you do this, make sure to switch it back later, we only want builds to occur on master, not on every branch.

CMCDragonkai commented:
We need `--no-bytecode` because without this flag it doesn't work inside 
Nix and it isn't deterministic. So up to your decision. As I said if you 
really need a directory, try `pkg` as the directory name to avoid confusion.

On 7/22/21 2:28 PM, Brian Botha wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In release.nix 
> <https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/29#discussion_r674493171>:
>
> > @@ -59,13 +65,16 @@ let
>         PKG_IGNORE_TAG = 1;
>         buildPhase = ''
>           cp ${./package.json} package.json
> +        mkdir -p output/prebuilds
>
> Ok, I can change it to pkg.
> as for |--public-packages "*" \| is just flagging everything is a 
> public package since non public packages are included as bytecode by 
> default.. This breaks with the |--no-bytecode| flag.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/29#discussion_r674493171>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAE4OHJ6LPSIJSE7VDU7XRLTY6M7PANCNFSM5AXJKZ6Q>.
>

CMCDragonkai commented:
You mean nixpkgs commit from master? Does this all work with just the 
revision provided by nixpkgs? 
https://github.com/NixOS/nixpkgs/commit/3d64b6067f1c457f889e91c59ced273be1d1e4f1

The above is the just the one that brought in `pkg`.

On 7/22/21 4:28 PM, Brian Botha wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In pkgs.nix 
> <https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/29#discussion_r674530748>:
>
> > @@ -1,5 +1,6 @@
>   import (
>     # let rev = "60cce7e5e1fdf62421ef6d4184ee399b46209366"; in
> -  let rev = "04a48a2fcd2370f4e8a8c9fd3394531ccdb30f1e"; in
> +  # let rev = "04a48a2fcd2370f4e8a8c9fd3394531ccdb30f1e"; in
> +  let rev = "53caacaf56640d04180775aee016d2f16d6f083c"; in
>
> I picked the latest pkgs commit when I started on this issue.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/29#discussion_r674530748>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAE4OHITJMVXE5N4SW42KT3TY63A5ANCNFSM5AXJKZ6Q>.
>

tegefaulkes commented:
I have created an upstram PR addressing the issue with the leveldown warnings.
vercel/pkg#1273

tegefaulkes commented:
When this is done, the changes for js-polykey are not far behind. the only complication would be the jose/jwt throwing warnings. But I've added them as an asset and it seems to be working with manual testing e.g. starting agent + starting a session.

CMCDragonkai commented:
I'm checking out your change to `vercel/pkg` now.

Firstly:

```
git fetch origin pull/1273/head:pr-1273
git checkout pr-1273
# setup nix-shell here
npm install
npm run build
# the build outputs to ./lib-es5, however the executable requires the node interpreter
node ./lib-es5/bin.js  --help
```

The output:

```

  pkg [options] <input>

  Options:

    -h, --help           output usage information
    -v, --version        output pkg version
    -t, --targets        comma-separated list of targets (see examples)
    -c, --config         package.json or any json file with top-level config
    --options            bake v8 options into executable to run with them on
    -o, --output         output file name or template for several files
    --out-path           path to save output one or more executables
    -d, --debug          show more information during packaging process [off]
    -b, --build          don't download prebuilt base binaries, build them
    --public             speed up and disclose the sources of top-level project
    --public-packages    force specified packages to be considered public
    --no-bytecode        skip bytecode generation and include source files as plain js
    -C, --compress       [default=None] compression algorithm = Brotli or GZip

  Examples:

  – Makes executables for Linux, macOS and Windows
    $ pkg index.js
  – Takes package.json from cwd and follows 'bin' entry
    $ pkg .
  – Makes executable for particular target machine
    $ pkg -t node14-win-arm64 index.js
  – Makes executables for target machines of your choice
    $ pkg -t node12-linux,node14-linux,node14-win index.js
  – Bakes '--expose-gc' and '--max-heap-size=34' into executable
    $ pkg --options "expose-gc,max-heap-size=34" index.js
  – Consider packageA and packageB to be public
    $ pkg --public-packages "packageA,packageB" index.js
  – Consider all packages to be public
    $ pkg --public-packages "*" index.js
  – Bakes '--expose-gc' into executable
    $ pkg --options expose-gc index.js
  – reduce size of the data packed inside the executable with GZip
    $ pkg --compress GZip index.js
```

I'm going to assume you manage to make it work.. so he main thing is that we need to get node2nix to work against your fork's branch.

CMCDragonkai commented:
@tegefaulkes In the future for this sort of work you would want to create the forks under MatrixAI group in github rather than your personal account: https://github.com/tegefaulkes/pkg

It makes it easier for us to maintain forks and maintain overrides in our nixpkgs overlay.

CMCDragonkai commented:
The other really important thing is to always branch off from `main` or `master` when creating a PR. This way you can have a `main` that syncs up with upstream, while creating a feature branch to be PRed into upstream main.

So for now I'll just create a feature branch off `main` and do the same as what you did.

CMCDragonkai commented:
This is now done here: https://github.com/MatrixAI/pkg/tree/leveldown

The steps were:

```
git remote add matrixai git@github.com:MatrixAI/pkg.git
git checkout main
git checkout -b leveldown
# make the change
git add -A
git commit -m 'Removes dictionary override for bundling leveldown - Fixes vercel#1264'
git push matrixai
```

CMCDragonkai commented:
The nix override work in our overlay is located here: https://gitlab.com/MatrixAI/Engineering/nixpkgs-overlay/-/merge_requests/15

tegefaulkes commented:
Ok, should be working now. Last thing is to check if the github CI still works.

CMCDragonkai commented on pkgs.nix:
```diff
@@ -1,5 +1,13 @@
+# Pinning to an overlay with a patch for vercel/pkg
+import (builtins.fetchGit {
+  url = "git@gitlab.com:MatrixAI/Engineering/nixpkgs-overlay.git";
+  rev = "958ab694978371a8d73c328cb515df85c7150960";
```
No need for this, just use the branch name below is sufficient. Plus I've rebased and squashed on the branch.

CMCDragonkai commented on pkgs.nix:
```diff
@@ -1,5 +1,13 @@
-  let rev = "04a48a2fcd2370f4e8a8c9fd3394531ccdb30f1e"; in
-  fetchTarball "https://github.com/MatrixAI/nixpkgs/archive/${rev}.tar.gz"
-)
+#import (
```
Remove these comments and just write a comment on this PR about how you came to using the relevant commits.

CMCDragonkai commented:
I don't think anything needs to change in our `.gitlab-ci.yml` file. It should work fine now.

So please fix up my comments, ensure any linting is done, and also clean up the README and format it and squash all the commits.

CMCDragonkai commented:
See comment here https://gitlab.com/MatrixAI/Engineering/nixpkgs-overlay/-/merge_requests/15/diffs#note_642632986 regarding our policy of using our private overlay in public packages.

tegefaulkes commented:
K, fixing up now.

CMCDragonkai commented:
After everything, don't forget to test out the Gitlab CI/CD with this method: https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/29#issuecomment-884637611

Then finally we can merge.

tegefaulkes commented:
I switched back to the public pkgs and added in the patch.
I've tested it in wsl and windows. It seems to be working. I'll try to run a pipeline in a second.

tegefaulkes commented:
Misclicked closed. My bad.

tegefaulkes commented:
Failing on tests. I'll look it over and fix it up tomorrow.

tegefaulkes commented:
Checks are passing, looks like it's working. this should be good for a merge after a review. @CMCDragonkai

tegefaulkes commented:
Squashed.

tegefaulkes commented:
That should be the last of it.
- [x] rebase
- [x] small fixes
- [x] squash
- [ ] review
- [ ] merge

CMCDragonkai commented:
Just fixed up a few things:

* Removed useless comments from pkgs.nix
* Formatted README.md with spacing between the code blocks and text
* Restoring the mocks should happen at the end of each test
* Indentation issues in `release.nix`

CMCDragonkai commented:
I noticed you also enabled source map for this repo. This is a good idea for a library. For a end-user application, not sure if it is relevant.

tegefaulkes commented:
> I noticed you also enabled source map for this repo. This is a good idea for a library. For a end-user application, not sure if it is relevant.

Ah, jest was throwing some minor errors without it. It's not needed I think.

CMCDragonkai commented:
You should also always test: `nix-build ./release.nix --attr application`.

Also we now have a custom override for our `pkg` here, this would have to done for all the repos using pkg.

Until upstream merges. Our overlay can also have this change applied too in the mean time.

CMCDragonkai commented:
The application build is currently failing due to the missing `node-gyp-build` command. It needs to be available to the derivation. We've dealt with this before.

CMCDragonkai commented:
So it turns out that what's required is that:

```
     node2nixProd = (
  -    (import (node2nixDrv false) { inherit pkgs nodejs; }).shell.override {
  +    (import (node2nixDrv false) { inherit pkgs nodejs; }).shell.override (attrs: {
  +      buildInputs = attrs.buildInputs ++ [ nodePackages.node-gyp-build ];
         dontNpmInstall = true;
  -    }
  +    })
```

The `override` can also take a function which takes the `attrs`, originally this was not the case. And there was a difference between `override` and `overrideAttrs`.

Interestingly you cannot make this work if you use `overrideAttrs`.

CMCDragonkai commented:
Now I find that you need `node-gyp-build` for both `node2nixProd` and `node2nixDev`.

This makes sense as the `node2nixProd` will only install production dependencies, whereas `node2nixDev` will install both production and development dependencies. So if there's an override, both require the `node-gyp-build` to be available.

Now that override and `overrideAttrs` both take a function with `attrs`, I wonder if this is because `override` delegates to `overrideAttrs` if it's an attribute function. Because nixpkgs upstream says that `override` is meant to override the function parameters. But this doesn't make any sense if there is access to existing attributes.

CMCDragonkai commented:
Yea I checked `override` being able to take `attrs` and applying something like `buildInputs` is specific to node2nix.

CMCDragonkai commented:
Regarding override issues: https://github.com/svanderburg/node2nix/issues/178

I've changed to using `override` instead of `overrideAttrs` for both `node2nixDev` and `node2nixProd`.

Regarding the build warnings of creating a homeless shelter that's cause `pkg` is attempting to download prebuilt-binaries but instead building from source should be possible.

This is solved with this environment variable in the release builds:

```
      # ensure that native modules are built from source
      npm_config_build_from_source = "true";
```

Regarding the application build, both dev and prod require the `node-gyp-build` due to leveldown. So this will be important for any future native modules we are adding in, the build system will become ever more complex. So it's ideal we document what each thing is for.

---

@tegefaulkes still has to fix up the executable itself, the arguments to `./result/bin/typescript-demo-lib` should be all optional. I think some arguments were introduced by @emmacasolin before where there was no defaulting value for them. Right now not passing it the right set of args results in weird errors.

This is also important for the docker output which doesn't pass any parameters to the command. It will look like something is broken without this being fixed.

CMCDragonkai commented:
My last commit ensures parity between nix-shell and nix-build. This should ensure that `pkg` being used inside `nix-shell` is the same as that which is used for the release. These changes will need to be applied to js-polykey too.

At the final point make sure to test all builds, application, docker, and the executables. And then merging time!
